### PR TITLE
Rdcc 5448 -Advance Search

### PR DESCRIPTION
**Pull request by sahitya-desireddy**
Advance Search API

JIRA link : https://tools.hmcts.net/jira/browse/RDCC-5448

**Does this PR introduce a breaking change?**

```
[ ] Yes
[X] No
```

### CODE REVIEW

**Review by shivajihmcts — CHANGES_REQUESTED** [review]
Can we convert it into dynamic query using Specification.

**Comment by sahitya-desireddy** [comment]
> Can we convert it into dynamic query using Specification.

Hi @shivajihmcts Thanks for your suggestion ,  we can go for specification when it is very complex query , but in advance search query we have to join multiple tables where it seems very complex if we use specification . Initially i thought of using specification but that will make more complex than the current used query so better to use the existing query rather than specification here.. else in case some one in the team can try this using specification we can take this up as an improvement in upcoming sprints :)

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+import static uk.gov.hmcts.reform.cwrdapi.TestSupport.validateSearchUserProfileResponse;
+import static uk.gov.hmcts.reform.cwrdapi.util.FeatureToggleConditionExtension.getToggledOffMessage;
+
+@ComponentScan("uk.gov.hmcts.reform.cwrdapi")
+@WithTags({@WithTag("testType:Functional")})
+@ActiveProfiles("functional")
+@SerenityTest
+@SpringBootTest
+@Slf4j
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
```
[the remark this answers is not in the document]
Comment (sahitya-desireddy) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java]: Done Thanks @SujeethKumarAri

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java [hunk 2]
```diff
@@ -323,4 +316,46 @@ public ResponseEntity<StaffProfileCreationResponse> createStaffUserProfile(@Requ
...
+    @ApiOperation(
+            value = "This API allows the Advance search of staff",
+            authorizations = {
+                    @Authorization(value = "ServiceAuthorization"),
+                    @Authorization(value = "Authorization")
+            }
+    )
+    @ApiResponses({
+            @ApiResponse(
+                    code = 201,
```
[the remark this answers is not in the document]
Comment (sahitya-desireddy) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java]: Modified Thank you @SujeethKumarAri

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/FeatureToggleServiceImpl.java [hunk 3]
```diff
@@ -57,6 +57,8 @@ public void mapServiceToFlag() {
                 CWD_FETCH_STAFF_BY_CCD_SERVICE_NAMES);
         launchDarklyMap.put("StaffRefDataController.searchStaffUserByName",
                 SRD_RD_STAFF_UI);
+        launchDarklyMap.put("StaffRefDataController.searchStaffProfile",
+                SRD_RD_STAFF_UI);
```
[the remark this answers is not in the document]
Comment (sahitya-desireddy) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/FeatureToggleServiceImpl.java:61]: Added  👍 @SujeethKumarAri

**Review by SujeethKumarAri — APPROVED** [review]

**Comment by sahitya-desireddy** [comment]
> Can we convert it into dynamic query using Specification.

**Review by shivajihmcts — APPROVED** [review]

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java [hunk 4]
```diff
@@ -86,6 +94,144 @@ public static List<CaseWorkerDomain> buildSuspendedCaseWorkerProfileData() {
...
+                    boolean valid = searchStaffUserResponse.getBaseLocations().stream()
+
+                            .anyMatch(location -> Optional.ofNullable(location).isPresent()
+                                    && finalSearchReq.getLocation().toLowerCase().contains(location.getBaseLocationId()
+                                    .toString().toLowerCase())
+                            );
+                    return valid;
+                })
+                .filter(searchStaffUserResponse -> {
+                    if (Optional.ofNullable(finalSearchReq.getLocation()).isEmpty()) {
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java:138]: duplicated , its already covered in above filter

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java [hunk 5]
```diff
@@ -86,6 +94,144 @@ public static List<CaseWorkerDomain> buildSuspendedCaseWorkerProfileData() {
                 .build());
     }
 
+
+    public static boolean validateSearchUserProfileResponse(ResponseEntity<List<SearchStaffUserResponse>> response,
+                                                            SearchRequest searchReq) {
+
+        SearchRequest finalSearchReq = searchReq;
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java:101]: local variable finalSearchReq is redundant

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java [hunk 6]
```diff
@@ -86,6 +94,144 @@ public static List<CaseWorkerDomain> buildSuspendedCaseWorkerProfileData() {
...
+                .filter(Objects::nonNull)
+                .filter(searchStaffUserResponse -> {
+                    if (Optional.ofNullable(finalSearchReq.getServiceCode()).isEmpty()) {
+                        return true;
+                    }
+                    if (Optional.ofNullable(searchStaffUserResponse.getServices()).isEmpty()) {
+                        return false;
+                    }
+
+                    boolean valid = searchStaffUserResponse.getServices().stream()
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java:114]: intellij warning inline variable boolean valid

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java [hunk 7]
```diff
@@ -86,6 +94,144 @@ public static List<CaseWorkerDomain> buildSuspendedCaseWorkerProfileData() {
...
+                })
+                .filter(searchStaffUserResponse -> {
+                    if (Optional.ofNullable(finalSearchReq.getLocation()).isEmpty()) {
+                        return true;
+                    }
+                    if (Optional.ofNullable(searchStaffUserResponse.getBaseLocations()).isEmpty()) {
+                        return false;
+                    }
+
+                    boolean valid = searchStaffUserResponse.getBaseLocations().stream()
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java:129]: intellij warning inline variable boolean valid

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java [hunk 8]
```diff
@@ -86,6 +94,144 @@ public static List<CaseWorkerDomain> buildSuspendedCaseWorkerProfileData() {
...
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java:145]: intellij warning inline variable boolean valid

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java [hunk 9]
```diff
@@ -86,6 +94,144 @@ public static List<CaseWorkerDomain> buildSuspendedCaseWorkerProfileData() {
...
+                                    && finalSearchReq.getLocation().toLowerCase().contains(String.valueOf(location
+                                    .getBaseLocationId()))
+                            );
+                    return valid;
+                })
+                .filter(searchStaffUserResponse -> {
+                    if (Optional.ofNullable(finalSearchReq.getUserType()).isEmpty()) {
+                        return true;
+                    }
+                    if (Optional.ofNullable(searchStaffUserResponse.getUserType()).isEmpty()) {
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java:157]: simplify if else

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java [hunk 10]
```diff
@@ -86,6 +94,144 @@ public static List<CaseWorkerDomain> buildSuspendedCaseWorkerProfileData() {
...
+                    return valid;
+                })
+                .filter(searchStaffUserResponse -> {
+                    if (Optional.ofNullable(finalSearchReq.getSkill()).isEmpty()) {
+                        return true;
+                    }
+                    if (Optional.ofNullable(searchStaffUserResponse.getSkills()).isEmpty()) {
+                        return false;
+                    }
+                    boolean valid = searchStaffUserResponse.getSkills().stream()
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java:184]: intellij warning inline variable boolean valid

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java [hunk 11]
```diff
@@ -86,6 +94,144 @@ public static List<CaseWorkerDomain> buildSuspendedCaseWorkerProfileData() {
...
+                            );
+                    return valid;
+                })
+                .filter(searchStaffUserResponse -> {
+                    if (Optional.ofNullable(finalSearchReq.getRole()).isEmpty()) {
+                        return true;
+                    }
+                    boolean valid = false;
+
+                    if (finalSearchReq.getRole().contains("task supervisor") && searchStaffUserResponse
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java:197]: check intellij warning

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java [hunk 12]
```diff
@@ -86,6 +94,144 @@ public static List<CaseWorkerDomain> buildSuspendedCaseWorkerProfileData() {
...
+                })
+
+
+                .toList();
+
+        Comparator<SearchStaffUserResponse> comparator
+                = Comparator.comparing(SearchStaffUserResponse::getLastName);
+
+        List<SearchStaffUserResponse> sorted = new ArrayList<>();
+        sorted.addAll(body);
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java:221]: check intellij warning. 'addAll()' call can be replaced with parametrized constructor call

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java [hunk 13]
```diff
@@ -86,6 +94,144 @@ public static List<CaseWorkerDomain> buildSuspendedCaseWorkerProfileData() {
...
+        Collections.sort(sorted,comparator);
+        if (!body.equals(sorted)) {
+            return false;
+        }
+
+        if (validResponse.size() == body.size()) {
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java:227]: 'if' statement can be simplified

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java [hunk 14]
```diff
@@ -86,6 +94,144 @@ public static List<CaseWorkerDomain> buildSuspendedCaseWorkerProfileData() {
                 .build());
     }
 
+
+    public static boolean validateSearchUserProfileResponse(ResponseEntity<List<SearchStaffUserResponse>> response,
+                                                            SearchRequest searchReq) {
+
+        SearchRequest finalSearchReq = searchReq;
+        List validResponse = new ArrayList<>();
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java:102]: change it to List<SearchStaffUserResponse>

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java [hunk 15]
```diff
@@ -336,6 +342,68 @@ void should_return_empty_list_of_case_worker_profile_with_status_code_200() {
...
+        when(caseWorkerProfileRepository.findByCaseWorkerProfiles(any(), any(), any(),
+                any(), any(), any(),any()))
+                .thenReturn(pages);
+        ResponseEntity<List<SearchStaffUserResponse>> responseEntity =
+                staffRefDataServiceImpl.retrieveStaffProfile(searchReq, pageRequest);
+        assertEquals(200, responseEntity.getStatusCodeValue());
+
+        assertThat(responseEntity.getHeaders().get("total-records").get(0)).isEqualTo("0");
+
+        List<SearchStaffUserResponse> searchResponse =
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java:400]: unused variable searchResponse

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java [hunk 16]
```diff
@@ -152,6 +217,379 @@ void shouldFetchEmptyUserTypes() {
...
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"abcde,fjkui","*__"})
+    void should_return_staffProfiles_search_with_invalid_input_location_status_code_400(String location) {
+        searchReq = SearchRequest.builder()
+                .location(location)
+                .build();
+
+        assertThrows(InvalidRequestException.class, () -> staffRefDataController
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java:322]: can you check for expected error message

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java [hunk 17]
```diff
@@ -152,6 +217,379 @@ void shouldFetchEmptyUserTypes() {
...
+
+    @ParameterizedTest
+    @ValueSource(strings = {"abc","*__","sdf sdf"})
+    void should_return_staffProfiles_search_with_invalid_input_usertype_status_code_400(String userType) {
+
+        searchReq = SearchRequest.builder()
+                .userType(userType)
+                .build();
+
+        assertThrows(InvalidRequestException.class, () -> staffRefDataController
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java:358]: can you check for expected error message

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java [hunk 18]
```diff
@@ -152,6 +217,379 @@ void shouldFetchEmptyUserTypes() {
...
+        assertTrue(validateSearchUserProfileResponse(response,searchReq));
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"s*_"})
+    void should_return_staffProfiles_search_with_invalid_service_code_400(String serviceId) {
+        searchReq = SearchRequest.builder()
+                .serviceCode(serviceId)
+                .build();
+        assertThrows(InvalidRequestException.class, () -> staffRefDataController
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java:251]: can you check for expected error message

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java [hunk 19]
```diff
@@ -152,6 +217,379 @@ void shouldFetchEmptyUserTypes() {
...
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"avdx","_*kd"})
+    void should_return_staffProfiles_search_with_invalid_input_jobTitle_status_code_400(String jobTitileId) {
+        searchReq = SearchRequest.builder()
+                .jobTitle(jobTitileId)
+                .build();
+
+        assertThrows(InvalidRequestException.class, () -> staffRefDataController
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java:395]: can you check for expected error message

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java [hunk 20]
```diff
@@ -152,6 +217,379 @@ void shouldFetchEmptyUserTypes() {
...
+
+    @ParameterizedTest
+    @ValueSource(strings = {"tasksupervisor,case_allocator,abcd"})
+    void should_return_staffProfiles_search_with_invalid_role_status_code_400(String roleId) {
+
+        searchReq = SearchRequest.builder()
+                .role(roleId)
+                .build();
+
+        assertThrows(InvalidRequestException.class, () -> staffRefDataController
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java:472]: can you check for expected error message

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java [hunk 21]
```diff
@@ -0,0 +1,464 @@
...
+                .getMultipleAuthHeadersWithoutContentTypeWithPagination(ROLE_STAFF_ADMIN,"1","1")
+                .get(CASE_WORKER_PROFILE_URL + "/search?" + searchString)
+                .andReturn();
+        fetchResponse.then()
+                .assertThat()
+                .statusCode(200);
+
+        int totalRecords = Integer.valueOf(fetchResponse.getHeader("total-records"));
+        if (totalRecords > 0) {
+            List<SearchStaffUserResponse> searchStaffUserResponse = Arrays.asList(
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java:151]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java [hunk 22]
```diff
@@ -336,6 +342,68 @@ void should_return_empty_list_of_case_worker_profile_with_status_code_200() {
...
+        var pageRequest = validateAndBuildPagination(20, 1,
+                20, 1);
+
+        ResponseEntity<List<SearchStaffUserResponse>> responseEntity =
+                staffRefDataServiceImpl.retrieveStaffProfile(searchReq, pageRequest);
+        assertEquals(200, responseEntity.getStatusCodeValue());
+
+        assertThat(responseEntity.getHeaders().get("total-records").get(0)).isEqualTo("1");
+
+        List<SearchStaffUserResponse> searchResponse =
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java:373]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java (hunk 15 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java:400]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java [hunk 23]
```diff
@@ -441,6 +509,7 @@ CaseWorkerProfile buildCaseWorkerProfile() {
         caseWorkerLocation.setPrimaryFlag(true);
 
         UserType userType = new UserType();
+        userType.setUserTypeId(1L);
         userType.setDescription("userTypeId");
 
         CaseWorkerProfile caseWorkerProfile = new CaseWorkerProfile();
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java:515]: changes done

**Comment by vasthav-patta** [comment]
@rajkatla-hmcts changes are made as per review comments

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java [hunk 24]
```diff
@@ -0,0 +1,464 @@
...
+                    fetchResponse.getBody().as(SearchStaffUserResponse[].class));
+            assertThat(searchStaffUserResponse).isNotNull();
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java]: done

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java [hunk 25]
```diff
@@ -0,0 +1,464 @@
...
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java]: done

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java [hunk 26]
```diff
@@ -0,0 +1,464 @@
...
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java]: done

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java [hunk 27]
```diff
@@ -0,0 +1,464 @@
...
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java]: done

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java [hunk 28]
```diff
@@ -0,0 +1,464 @@
...
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefDataAdvancedSearchFunctionalTest.java]: done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java [hunk 29]
```diff
@@ -152,6 +217,379 @@ void shouldFetchEmptyUserTypes() {
...
+        verify(staffRefDataService, times(0))
+                .retrieveStaffProfile(eq(searchReq), Mockito.any(PageRequest.class));
+
+    }
+
+
+    @Test
+    void should_return_staffProfiles_with_status_code_400_at_least_one_param_required() {
+        SearchRequest req = SearchRequest.builder().build();
+        assertThrows(EmptyRequestException.class, () -> staffRefDataController
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java [hunk 30]
```diff
@@ -152,6 +217,379 @@ void shouldFetchEmptyUserTypes() {
...
+        assertThrows(InvalidRequestException.class, () -> staffRefDataController
+                .searchStaffProfile(0,2, searchReq));
+        verify(staffRefDataService, times(0))
+                .retrieveStaffProfile(eq(searchReq), Mockito.any(PageRequest.class));
+
+    }
+
+    @Test
+    void should_return_staffProfiles_with_status_code_400_page_size_lessThan1() {
+        assertThrows(InvalidRequestException.class, () -> staffRefDataController
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java [hunk 31]
```diff
@@ -152,6 +217,379 @@ void shouldFetchEmptyUserTypes() {
...
+        int totalRecords = Integer.valueOf(response.getHeaders().get("total-records").get(0));
+        assertThat(totalRecords).isEqualTo(2);
+        verify(staffRefDataService, times(1))
+                .retrieveStaffProfile(eq(searchReq), Mockito.any(PageRequest.class));
+    }
+
+
+    @Test
+    void should_return_staffProfiles_with_status_code_400_page_number_0() {
+        assertThrows(InvalidRequestException.class, () -> staffRefDataController
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java (hunk 20 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java [hunk 32]
```diff
@@ -152,6 +217,379 @@ void shouldFetchEmptyUserTypes() {
...
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"avdx","_*kd"})
+    void should_return_staffProfiles_search_with_invalid_skillId_status_code_400(String skillId) {
+        searchReq = SearchRequest.builder()
+                .skill(skillId)
+                .build();
+
+        assertThrows(InvalidRequestException.class, () -> staffRefDataController
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java (hunk 19 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java (hunk 17 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java (hunk 16 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java (hunk 18 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java (hunk 13 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java [hunk 33]
```diff
@@ -86,6 +94,144 @@ public static List<CaseWorkerDomain> buildSuspendedCaseWorkerProfileData() {
...
+
+
+                .toList();
+
+        Comparator<SearchStaffUserResponse> comparator
+                = Comparator.comparing(SearchStaffUserResponse::getLastName);
+
+        List<SearchStaffUserResponse> sorted = new ArrayList<>();
+        sorted.addAll(body);
+        Collections.sort(sorted,comparator);
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java (hunk 12 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java (hunk 11 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java (hunk 10 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java (hunk 9 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java (hunk 8 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java (hunk 4 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java (hunk 7 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java (hunk 6 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java (hunk 14 above)
Comment (vasthav-patta) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/TestSupport.java]: changes done

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java (hunk 15 above)
Comment (rajkatla-hmcts) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceImplTest.java]: @vasthav-patta , i can still see the unused variable searchResponse, pls check

**Review by rajkatla-hmcts — APPROVED** [review]

File: src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/StaffReferenceProfileAdvanceSearchTest.java [hunk 34]
```diff
@@ -0,0 +1,588 @@
...
+                .searchStaffUser(path, searchString, "1", "1", ROLE_STAFF_ADMIN);
+        assertThat(response).containsEntry("http_status", "400");
+        assertThat(response.get("response_body").toString())
+                .contains("Bad Request");
+
+    }
+
+
+
+    void createCaseWorkerTestData(String firstName, String lastName, String email) {
```
[the remark this answers is not in the document]
Comment (vasthav-patta) [reply src/integrationTest/java/uk/gov/hmcts/reform/cwrdapi/StaffReferenceProfileAdvanceSearchTest.java]: done
https://tools.hmcts.net/jira/browse/RDCC-5758

**Pull request merged**